File: 백준/Silver/10844. 쉬운 계단 수/쉬운 계단 수.cc
#include <iostream>
#include <vector>
using namespace std;

static long long int MOD = 1000000000;
long long int N;
vector<vector<long long int>> stairNumber(10, vector<long long int>(101, -1));

int getStairNumber(int digit, int len) {
	if (digit < 0 || digit > 9) return 0;
	int result = stairNumber[digit][len];
	if (result != -1) return result;
	else {
		stairNumber[digit][len] = (getStairNumber(digit - 1, len - 1) + getStairNumber(digit + 1, len - 1)) % MOD;
		return stairNumber[digit][len];
	}
}

int main() {
	ios_base::sync_with_stdio(0);
	cin.tie(0);
	cout.tie(0);

	cin >> N;
	int result = 0;
	for (int i = 0; i < 10; i++) {
		stairNumber[i][1] = 1;
	}
	for (int i = 1; i < 10; i++) {
		result += getStairNumber(i, N);
		result %= MOD;
	}
	cout << result % MOD << "\n";

	return 0;
}

File: 백준/Silver/10844. 쉬운 계단 수/쉬운 계단 수.js
const fs = require("fs");
const input = fs
  .readFileSync(
    process.platform === "linux" ? "/dev/stdin" : "input.txt",
    "utf8"
  )
  .trim()
  .split("\n");

const N = +input[0];
const MOD = 1000000000;

const getStairNumber = (digit, len, stairNumber) => {
  if (digit < 0 || digit > 9) return 0;
  else if (stairNumber[digit][len] !== undefined)
    return stairNumber[digit][len];
  else {
    stairNumber[digit][len] =
      (getStairNumber(digit - 1, len - 1, stairNumber) +
        getStairNumber(digit + 1, len - 1, stairNumber)) %
      MOD;
    return stairNumber[digit][len];
  }
};

const solution = (N) => {
  const stairNumber = Array.from({ length: 10 }, () => []);
  for (let i = 0; i <= 9; i++) stairNumber[i][1] = 1;

  let result = 0;
  for (let i = 1; i <= 9; i++)
    result = (result + getStairNumber(i, N, stairNumber)) % MOD;
  return result;
};

console.log(solution(N));
